content/blog/2024-08-29-valkey-memory-efficiency-8-0.md

```diff
@@ -0,0 +1,148 @@
+
+![Figure 4 Dictionary in cluster mode (Valkey 8.0) with multiple key value pair](/assets/media/pictures/dictionary_in_ cluster_mode_8.0.png)
+
+Few of the interesting challenges that comes up with the above improvements are supporting existing use cases which were optimized with a single dictionary for the entire keyspace. The usecases are:
+
+* [Iterating the entire keyspace](https://valkey.io/commands/scan/) - Command like `SCAN` to iterate over the entire keyspace.
+* [Random key for eviction](https://valkey.io/topics/lru-cache/) - The server does random sampling of the keyspace to find ideal candidate for eviction.
+* [Finding a random key](https://valkey.io/commands/randomkey/) - Command like `RANDOMKEY` retrieves a random key from the database.
+
+
+It is quite possible that some slots contains more keys than others and some could be also empty which could lead the engine to traverse through the all the empty dictionary for a full scan i.e. [O(N)](https://en.wikipedia.org/wiki/Big_O_notation) operation if approached naively. When picking a random key for eviction or running `RANDOMKEY` operation, the server needs to pick a non-empty slot in a performant manner. These cases require a data structure which provides the following functionality:
+
+1. Modify value for a given index (slot) - If a key gets added or removed, increment or decrement the value for that given slot by 1 respectively.
+2. Cumulative frequency until each index (slot) - For a given number representing a key between 1 and the total number of keys, return the index (slot) which covers the particular key.
+
+If approached naively, the former and latter operation would take O(1) and O(N) respectively. However, we want to minimize the latter operation’s time complexity and avoid regression in the operations stated above. Hence, a [binary indexed tree (BIT) or fenwick tree](https://www.topcoder.com/thrive/articles/Binary%20Indexed%20Trees) which provides the above functionality with a minimal memory overhead (~1 MB per node) and the time complexity is also bounded to O(M log N) for both the operations where M = number of modification(s) and N = number of slots. This enables skipping over empty slot efficiently while iterating over the keyspace as well as finding a slot for a given key index in logarithmic time via binary search over the cummulative sum maintained by the BIT. 
```

Comment:
`This enables skipping over empty slot efficiently` -> `This enables skipping over empty slots efficiently`